text/0000-iri-v181-transaction-and-bundle/0000-iri-v181-transaction-and-bundle.md

```diff
@@ -0,0 +1,980 @@
+implementation. Conversion methods shall only verify that the passed byte slices (or fixed-size arrays or vectors of
+bytes) are of appropriate length and that each contained byte correctly encodes a balanced trit (this is also referred
+to as `T1B1` binary-coded ternary encoding).
+
+```rust
+pub struct SignatureOrMessageFragment([u8; 6561]);
+pub struct Address([u8; 243]);
+pub struct Value([u8; 81]);
+pub struct Tag([u8; 81]);
+pub struct Timestamp([u8; 27]);
+pub struct Index([u8; 27]);
+pub struct BundleHash([u8; 243]);
+pub struct TransactionHash([u8; 243]);
+pub struct Nonce([u8; 81]);
+```
+
+Below is an example implementation for the `Tag`, with conversion from a `&[u8]`. An implementation in terms of `&[i8]`
+would look similar. Checked conversions ensuring that each byte encodes `-1`, `0`, or `+1` are implemented in terms of
+the `TryFrom` trait, while a `Tag::from_unchecked` function is provided as an unchecked faster constructor method.
+
+```rust
+enum TagError {
+    InvalidBinaryEncodedTrit,
+    WrongLength { expected: usize, given: usize },
+}
+
+impl fmt::Display for TagError {
+    fn display(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        use TagError::*;
+
+        let msg: Cow<'_, str> = match self {
+            InvalidBinaryEncodedTrit => "Failed to parse u8; only 0b0000_0000, 0b0000_0001, 0b1111_1111 are valid values.".into(),
+
+            WrongLength { expected, given } =>
+                format!("Expected {} characters, found {} in String").into(),
+        };
+
+        fmt!(f, &msg)
+    }
+}
+
+impl error::Error for TagError {
+    fn source(&self) -> Option<&(dyn error::Error + 'static)> {
+        None
+    }
+}
+
+impl Tag {
+    /// Unchecked conversion from a byte slice.
+    ///
+    /// Panics if the slice has the wrong length, and does not check if the bytes correctly encode
+    /// a trit.
+    fn from_unchecked(bytes: &[u8]) -> Self {
+        let mut array = [0u8; 81];
+        // Panics if `bytes` is not of length 81
+        Tag(array.copy_from_slice(bytes))
+    }
+}
+
+impl<'_> TryFrom<&'_ [u8]> for Tag {
+    type Error = TagError;
+
+    fn try_from(value: &'_ [u8]) -> Result<Self, Self::Error> {
+        if value.len() != 81 {
+            Err(TagError::WrongLength { expected: 81, given: value.len()})?;
+        }
+
+        for byte in value {
+            match byte {
+                0b0000_0000 | 0b1111_1111 | 0b0000_0001 => {},
+                _ => Err(TagError::InvalidBinaryEncodedTrit)?,
+        }
+
+        Ok(Tag::from_unchecked(value))
+    }
+}
+```
+
+### `TransactionBuilder` struct
+
+The `TransactionBuilder` allows setting all the fields of a transaction, and verifies and builds a correct `Transaction`
+type. The order of the fields in `TransactionBuilder` follows the table showing the different fields of a transaction,
+the same as `Transaction`.
+
+```rust
+pub struct TransactionBuilder {
```

Comment:
Just a little thing here, but the TransactionBuilder actually doesn't build the Transaction, it's the BundleBuilder which fills indices, bundle hash, etc. and finally builds it right?
So maybe we should call it `TransctionDraft` instead of `TransactionBuilder`, as you suggested before
Letting the BundleBuilder then direclty operate on the `TransctionDraft` (with all fields there: bundlehash, indices, timestamps, ... ) seems clean to me :+1: